cpp/src/arrow/allocator.h

```diff
@@ -85,6 +85,41 @@ class stl_allocator {
   MemoryPool* pool_;
 };
 
+template <typename Allocator = std::allocator<uint8_t>>
+class STLMemoryPool : public MemoryPool {
+ public:
+  explicit STLMemoryPool(const Allocator& alloc) : alloc_(alloc) {}
+  ~STLMemoryPool() override = default;
+
+  Status Allocate(int64_t size, uint8_t** out) override {
+    *out = alloc_.allocate(size);
+    stats_.UpdateAllocatedBytes(size);
+    return Status::OK();
+  }
+
+  Status Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) override {
+    uint8_t* old_ptr = *ptr;
+    *ptr = alloc_.allocate(new_size);
+    memcpy(*ptr, old_ptr, std::min(old_size, new_size));
+    Free(old_ptr, old_size);
```

Comment:
This will wrongly call `UpdateAllocatedBytes`.

Reply:
Let's say that before reallocate, you have `bytes_allocated = old_size`.
After reallocate, you want to have `bytes_allocated = new_size`.
But in reality, you have:
* after the call to `Free(old_ptr, old_size)`: `bytes_allocated = 0`
* after the call to `UpdateAllocatedBytes(new_size - old_size)`: `bytes_allocated = new_size - old-size`.

And this is visible in the test results btw: https://travis-ci.org/apache/arrow/jobs/447042467#L1094

Reply:
🤦‍♂️ I was running the release tests while having a debug build. That explains my confusion.

Reply:
Fixed now